src/track/globaltrackcache.cpp

```diff
@@ -300,296 +422,309 @@ TrackPointer GlobalTrackCache::lookupInternal(
+            // Due to a rare but expected race condition the track
+            // has already been deleted and must not be deleted
+            // again!
+            if (debugLogEnabled()) {
+                kLogger.debug()
+                        << "Skip deletion of already deleted track"
+                        << plainPtr;
             }
+            return false;
         }
-        m_tracksByCanonicalLocation.erase(trackByCanonicalLocation);
-        Stat::track(kEraseByCanonicalLocationCounter, Stat::COUNTER, kStatCounterFlags, 1);
     }
-    return purgedItem;
+
+    // Now we know that the pointer has not been deleted before
+    // and we can safely access it!
+    return evictAndDelete(
+            &cacheLocker,
+            indexedTrack,
+            false);
 }
 
-Track* GlobalTrackCache::evictInternal(
+bool GlobalTrackCache::evictAndDelete(
         GlobalTrackCacheLocker* pCacheLocker,
-        const TrackRef& trackRef) {
-    if (kLogger.debugEnabled()) {
+        IndexedTracks::iterator indexedTrack,
+        bool evictUnexpired) {
+    Track* plainPtr = (*indexedTrack).first;
+    DEBUG_ASSERT(plainPtr);
+    DEBUG_ASSERT(m_unindexedTracks.find(plainPtr) == m_unindexedTracks.end());
+    const auto trackRef = createTrackRef(*plainPtr);
+    if (debugLogEnabled()) {
         kLogger.debug()
-                << "Evicting track"
-                << trackRef;
+                << "Evicting indexed track"
+                << trackRef
+                << plainPtr;
     }
 
-    const Item purgedItem = purgeInternal(trackRef);
+    const bool evicted = evict(trackRef, indexedTrack, evictUnexpired);
     DEBUG_ASSERT(verifyConsistency());
-    if (nullptr != purgedItem.plainPtr) {
-        // It can produce dangerous signal loops if the track is still
-        // sending signals while being saved! All references to this
-        // track have been dropped at this point, so there is no need
-        // to send any signals.
-        // See: https://bugs.launchpad.net/mixxx/+bug/1365708
-        purgedItem.plainPtr->blockSignals(true);
-
-        // Keep the cache locked while evicting the track object!
-        // The callback is given the chance to unlock the cache
-        // after all operations that rely on managed track ownership
-        // have been done, e.g. exporting track metadata into a file.
-        m_pEvictor->onEvictingTrackFromCache(
-                pCacheLocker,
-                purgedItem.plainPtr);
-        Stat::track(kEvictCounter, Stat::COUNTER, kStatCounterFlags, 1);
-        // At this point the cache might have been unlocked already
+    if (evicted) {
+        // The evicted entry must not be accessible anymore!
+        DEBUG_ASSERT(m_indexedTracks.find(plainPtr) == m_indexedTracks.end());
+        DEBUG_ASSERT(!lookupByRef(trackRef));
+        m_pDeleter->onDeleteTrackBeforeUnlockingCache(plainPtr);
+        if (pCacheLocker) {
+            pCacheLocker->unlockCache();
+        }
```

Comment:
Here, a track object might be created from a different thread. It will load the old metadata from the library.
The later save, will revert the already saved new data.

Reply:
How to solve this? The issue is hat we cannot hold the lock due to the risk of deadlocks.

Reply:
This race condition has not been introduced by the cache. It was already there in previous versions with completely unmanaged track object.

Race conditions are impossible if the deletion is invoked from the main thread. This is currently the only case as my debug assertions have proven. I agree that this assumption might not hold forever.

If the deletion is invoked from another thread keeping the cache locked doesn't help either. In this case the command for updating the database is sent through a queued connection to the main thread and handled when dispatched by its event loop. Keeping the cache locked her while just enqueueing this message does not solve any race conditions.

When enabling multi-threaded database access sometimes in the future we don't need the queued connection any more and can update the database directly from the deleting thread. In this case we have the choice between safety (keeping the cache locked while writing to the database) vs. responsiveness (unlocking the cache before starting a database write). If we want to ensure to take the safe path we can already do everything within the onDeleteTrackBeforeUnlockingCache() operation. But in the current situation this doesn't provide any additional safety guarantees and hides a race condition!

If you prefer to perform all actions within the locking scope I will change it. But we need to explain that and why this currently doesn't prevent any race conditions between database updates!

Reply:
Ok, the code can be simplified with only a single callback while keeping the cache locked. With some appropriate comments this should pave the way towards multi-threaded database access.